nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java

```diff
@@ -0,0 +1,382 @@
+        .name("FAILURE")
+        .description("All FlowFiles that failed to be written to MarkLogic are routed to the " +
+            "failure relationship for future processing")
+        .build();
+
+    private DataMovementManager dataMovementManager;
+    private WriteBatcher writeBatcher;
+    // If no FlowFile exists when this processor is triggered, this variable determines whether or not a call is made to
+    // flush the WriteBatcher
+    private boolean shouldFlushIfEmpty = true;
+
+    @Override
+    public void init(ProcessorInitializationContext context) {
+        super.init(context);
+
+        List<PropertyDescriptor> list = new ArrayList<>();
+        list.addAll(properties);
+        list.add(COLLECTIONS);
+        list.add(FORMAT);
+        list.add(JOB_ID);
+        list.add(JOB_NAME);
+        list.add(MIMETYPE);
+        list.add(PERMISSIONS);
+        list.add(TRANSFORM);
+        list.add(TEMPORAL_COLLECTION);
+        list.add(URI_ATTRIBUTE_NAME);
+        list.add(URI_PREFIX);
+        list.add(URI_SUFFIX);
+        properties = Collections.unmodifiableList(list);
+        Set<Relationship> set = new HashSet<>();
+        set.add(SUCCESS);
+        set.add(FAILURE);
+        relationships = Collections.unmodifiableSet(set);
+    }
+
+    @OnScheduled
+    public void onScheduled(ProcessContext context) {
+        dataMovementManager = getDatabaseClient(context).newDataMovementManager();
+        writeBatcher = dataMovementManager.newWriteBatcher()
+            .withJobId(context.getProperty(JOB_ID).getValue())
+            .withJobName(context.getProperty(JOB_NAME).getValue())
+            .withBatchSize(context.getProperty(BATCH_SIZE).asInteger())
+            .withThreadCount(context.getProperty(THREAD_COUNT).asInteger())
+            .withTemporalCollection(context.getProperty(TEMPORAL_COLLECTION).getValue());
+
+        final String transform = context.getProperty(TRANSFORM).getValue();
+        if (transform != null) {
+            writeBatcher.withTransform(new ServerTransform(transform));
+        }
+        this.writeBatcher.onBatchSuccess(writeBatch -> {
+            for(WriteEvent writeEvent : writeBatch.getItems()) {
+                routeDocumentToRelationship(writeEvent, SUCCESS);
+            }
+        }).onBatchFailure((writeBatch, throwable) -> {
+            for(WriteEvent writeEvent : writeBatch.getItems()) {
+                routeDocumentToRelationship(writeEvent, FAILURE);
+            }
+        });
+        dataMovementManager.startJob(writeBatcher);
+    }
+
+    private void routeDocumentToRelationship(WriteEvent writeEvent, Relationship relationship) {
+        DocumentMetadataHandle metadata = (DocumentMetadataHandle) writeEvent.getMetadata();
+        String flowFileUUID = metadata.getMetadataValues().get("flowFileUUID");
+        FlowFileInfo flowFile = URIFlowFileMap.get(flowFileUUID);
+        if(flowFile != null) {
```

Comment:
Have you tried this with a very large data set to see how it performs? Like millions of sample XML records?